Make Attributes Immutable #80

TysonAndre opened the pull request:
Finish making a lot of code in the Intermediate immutable, including
Attributes.

Fixing tests for this.

This does not include type checking of attributes.

This does not yet distinguish between an attribute that is missing and
attribute that is set but not expected. The number of exceptions to be
listed in "throws" would grow large, and may need to be rethought.

TysonAndre committed:
Make Attributes Immutable #80

Finish making a lot of code in the Intermediate immutable, including
Attributes.

Fixing tests for this.

This does not include type checking of attributes.

This does not yet distinguish between an attribute that is missing and
attribute that is set but not expected. The number of exceptions to be
listed in "throws" would grow large, and may need to be rethought.

mtrberzi commented:
Checkstyle is preventing the build from going through, have a look at the style report and push again.

mtrberzi commented:
References #80

mtrberzi commented on src/main/java/org/manifold/intermediate/Attributes.java:
```diff
@@ -1,13 +1,28 @@
+    }
+    for (String attrName: data.keySet()) {
+      if (!types.containsKey(attrName)) {
+        // TODO: Should this be a different type, or should there be a common base class?
```
I think it's okay for this to be a different exception type. There really are quite a few ways that attributes can "go wrong" from the perspective of bad typing, and I think it's more useful to have separate exceptions for each condition to be able to indicate what has gone wrong as precisely as possible.

lucaswoj replied:
:thumbsup:

mtrberzi commented:
For the most part this seems okay. Just look over the code style nits given by checkstyle and I think this is ready to go.

lucaswoj commented on src/main/java/org/manifold/intermediate/Attributes.java:
```diff
@@ -1,13 +1,28 @@
 
-  public Attributes() {
-    data = new HashMap<>();
+  public Attributes(Map<String, Type> types, Map<String, Value> data)
```
Do you think we eventually will need a class to encapsulate `AttributeTypes` (ie the `Map<String, Type> types`) above? (This would be a new PR)

mtrberzi replied:
Hmm, I really don't see that being anything more than a wrapper for this `Map`...

lucaswoj replied:
True. I guess the prudent thing is to wait until it becomes a problem!

lucaswoj commented:
You can see the checkstyle output inline in your IDE or on the command line after running `gradle check`. Travis runs `gradle check`. Here are the checkstyle violations for this commit:

```
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Attributes.java:0: File does not end with a newline.
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Attributes.java:5: warning: Wrong order for 'com.google.common.collect.ImmutableMap' import.
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Attributes.java:12:39: ':' is not preceded with whitespace.
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Attributes.java:19:25: ':' is not preceded with whitespace.
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Attributes.java:21: Line is longer than 80 characters (found 89).
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Connection.java:26: Line is longer than 80 characters (found 87).
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/ConnectionType.java:5: warning: Wrong order for 'com.google.common.collect.ImmutableMap' import.
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Constraint.java:14: Line is longer than 80 characters (found 104).
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/ConstraintType.java:5: warning: Wrong order for 'com.google.common.collect.ImmutableMap' import.
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Node.java:24: Line is longer than 80 characters (found 169).
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Node.java:31:27: ':' is not preceded with whitespace.
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Node.java:41: Line is longer than 80 characters (found 99).
[ant:checkstyle] /home/travis/build/manifold-lang/manifold/src/main/java/org/manifold/intermediate/Port.java:21: Line is longer than 80 characters (found 107).
```

I'm open to discussions about how to evolve our checkstyle rules to suit everyone's preferences. Off the bat, I'm not sure I really care about import ordering.

TysonAndre committed:
Fix main checkstyle

TysonAndre committed:
Add base class SchematicException and more types of exceptions

-	add base class for exception caused by an invalid schematic
-	add InvalidAttributeException and InvalidIdentifierException.
-	fix checkstyle errors in the tests directory
-	Make some tests expect SchematicException instead of a long list
-	remove TODOs

mtrberzi commented:
Thanks Tyson! This looks good. Merge it!
